lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -622,20 +622,31 @@ def add_mpi_data_for_network_tokenization_card(post, payment, options)
 
         post[:mpiData] = {}
         post[:mpiData][:authenticationResponse] = 'Y'
-        post[:mpiData][:cavv] = payment.payment_cryptogram
+        if NETWORK_TOKENIZATION_CARD_SOURCE[payment.source.to_s].nil? && options[:switch_cryptogram_mapping_nt]
+          post[:mpiData][:tokenAuthenticationVerificationValue] = payment.payment_cryptogram
+        else
+          post[:mpiData][:cavv] = payment.payment_cryptogram
+        end
         post[:mpiData][:directoryResponse] = 'Y'
         post[:mpiData][:eci] = payment.eci || '07'
       end
 
-      def add_recurring_contract(post, options = {})
-        return unless options[:recurring_contract_type]
+      def add_recurring_contract(post, options = {}, payment = nil)
+        return unless options[:recurring_contract_type] || (payment.is_a?(NetworkTokenizationCreditCard) && options[:switch_cryptogram_mapping_nt])
```

Comment:
NetworkTokenizationCreditCard is for ApplePay, GooglePay and NetworkToken and `payment.source` is what tells you what exactly it is. This is why because of this it's why here you have to updated it to be just like you did in [line 625](https://github.com/activemerchant/active_merchant/pull/5155/files#diff-022d82d7d09d8eb787e16961826d7f88f536d9eb2a9ef947432dcf26250258e0R625) or you can try `payment.try(:source) == :network_token`

Reply:
Updated!